src/AppInstallerCLICore/PortableInstaller.cpp

```diff
@@ -0,0 +1,460 @@
+                CommitToARPEntry(PortableValueName::SHA256, entry.SHA256);
+            }
+
+            Filesystem::RenameFile(entry.CurrentPath, filePath);
+        }
+        else if (fileType == PortableFileType::Directory)
+        {
+            AICLI_LOG(Core, Info, << "Moving directory to: " << filePath);
+            Filesystem::RenameFile(entry.CurrentPath, filePath);
+        }
+        else if (entry.FileType == PortableFileType::Symlink && !InstallDirectoryAddedToPath)
+        {
+            std::filesystem::file_status status = std::filesystem::status(filePath);
+            if (std::filesystem::is_directory(status))
+            {
+                AICLI_LOG(CLI, Info, << "Unable to create symlink. '" << filePath << "points to an existing directory.");
+                THROW_HR(APPINSTALLER_CLI_ERROR_PORTABLE_SYMLINK_PATH_IS_DIRECTORY);
+            }
+
+            if (!RecordToIndex)
+            {
+                CommitToARPEntry(PortableValueName::PortableSymlinkFullPath, filePath);
+            }
+
+            if (std::filesystem::remove(filePath))
+            {
+                AICLI_LOG(CLI, Info, << "Removed existing file at " << filePath);
+                m_stream << Resource::String::OverwritingExistingFileAtMessage << ' ' << filePath << std::endl;
+            }
+
+            if (Filesystem::CreateSymlink(entry.SymlinkTarget, filePath))
+            {
+                AICLI_LOG(Core, Info, << "Symlink created at: " << filePath);
+            }
+            else
+            {
+                // Symlink creation should only fail if the user executes without admin rights or developer mode.
+                // Resort to adding install directory to PATH directly.
+                AICLI_LOG(Core, Info, << "Portable install executed in user mode. Adding package directory to PATH.");
+                CommitToARPEntry(PortableValueName::InstallDirectoryAddedToPath, InstallDirectoryAddedToPath = true);
+            }
+        }
+    }
+
+    void PortableInstaller::RemoveFile(AppInstaller::Portable::PortableFileEntry& entry)
+    {
+        const auto& filePath = entry.GetFilePath();
+        PortableFileType fileType = entry.FileType;
+
+        if (fileType == PortableFileType::File && std::filesystem::exists(filePath))
+        {
+            AICLI_LOG(CLI, Info, << "Deleting portable exe at: " << filePath);
+            std::filesystem::remove(filePath);
+        }
+        else if (fileType == PortableFileType::Symlink && Filesystem::SymlinkExists(filePath))
+        {
+            AICLI_LOG(CLI, Info, << "Deleting portable symlink at: " << filePath);
+            std::filesystem::remove(filePath);
+        }
+        else if (fileType == PortableFileType::Directory && std::filesystem::exists(filePath))
+        {
+            AICLI_LOG(CLI, Info, << "Removing directory at " << filePath);
+            std::filesystem::remove_all(filePath);
+        }
+    }
+
+    void PortableInstaller::ApplyDesiredState()

```

Comment:
So while this has the structure of what I was pushing for (a single function that applies the state), this is a very inefficient implementation.  I was hoping that we would be able to apply the differences rather than removing everything and putting all the files back.  That isn't strictly necessary; we could have a follow-up PR at some point to handle that.  It is possible that we will find it necessary to do before shipping for real depending on the performance/size of the archives that people are using.

Additionally, while it also isn't necessary to do it this way, it would be cleaner if the index/no index was decided once by creating an object that implemented the necessary interface.  That way you just throw values at it and it decides where to put them, rather than having if blocks sprinkled throughout.